In [2]:
import pandas as pd

In [3]:
df = pd.read_sas('Data/nsch_2020_topical_SAS/nsch_2020_topical.sas7bdat')

/Users/matthewturner/opt/anaconda3/envs/Capstone/lib/python3.9/site-packages/pandas/io/sas/sas7bdat.py:800: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  rslt[name] = self._byte_chunk[jb, :].view(dtype=self.byte_order + "d")


In [4]:
df.shape

(42777, 443)

In [5]:
df.head()

FIPSST STRATUM         HHID FORMTYPE  TOTKIDS_R  TENURE  HHLANGUAGE  \
0  b'17'    b'1'  b'20000003'    b'T1'        2.0     1.0         1.0   
1  b'29'   b'2A'  b'20000004'    b'T3'        1.0     1.0         1.0   
2  b'47'    b'1'  b'20000005'    b'T1'        1.0     1.0         1.0   
3  b'28'    b'1'  b'20000014'    b'T3'        2.0     1.0         1.0   
4  b'55'    b'1'  b'20000015'    b'T3'        2.0     2.0         1.0   

   SC_AGE_YEARS  SC_SEX  K2Q35A_1_YEARS  ...  BIRTH_YR_F  BMICLASS  \
0           3.0     1.0             NaN  ...         0.0       NaN   
1          14.0     2.0             NaN  ...         0.0       2.0   
2           1.0     2.0             NaN  ...         0.0       NaN   
3          15.0     2.0             NaN  ...         0.0       2.0   
4          16.0     2.0             NaN  ...         0.0       3.0   

   HHCOUNT_IF  FPL_I1  FPL_I2  FPL_I3  FPL_I4  FPL_I5  FPL_I6          FWC  
0         0.0   400.0   400.0   400.0   400.0   400.0   400.0  3296.080092  
1         0.0   400.0   400.0   400.0   400.0   400.0   400.0  2888.545330  
2         0.0   400.0   400.0   400.0   400.0   400.0   400.0  1016.682730  
3         0.0   143.0   143.0   143.0   143.0   143.0   143.0  1042.091065  
4         0.0   400.0   400.0   400.0   400.0   400.0   400.0   402.372392  

[5 rows x 443 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42777 entries, 0 to 42776
Columns: 443 entries, FIPSST to FWC
dtypes: float64(439), object(4)
memory usage: 144.6+ MB


There are 439 float columns and "4" object columns.

In [7]:
# Header: Has a doctor or other health care provider EVER told you that this child has Attention Deficit Disorder or Attention-Deficit/Hyperactivity Disorder, that is, ADD or ADHD?
df['K2Q31A'].value_counts()

2.0    38283
1.0     4306
Name: K2Q31A, dtype: int64

Of the 42,777 kids, 4306 have been diagnosed with ADHD by a doctor.

In [8]:
# If yes, does this child CURRENTLY have the condition?
# 1 = yes, 2 = no
df['K2Q31B'].value_counts(normalize=True)

1.0    0.922038
2.0    0.077962
Name: K2Q31B, dtype: float64

of the 4306 diagnoses, 3891 CURRENTLY have the condition? This seems illogical at first glance, perhaps a misdiagnosis

In [9]:
# A baseline mode

In [10]:
# All Object type columns
df.select_dtypes('object')

FIPSST STRATUM         HHID FORMTYPE
0      b'17'    b'1'  b'20000003'    b'T1'
1      b'29'   b'2A'  b'20000004'    b'T3'
2      b'47'    b'1'  b'20000005'    b'T1'
3      b'28'    b'1'  b'20000014'    b'T3'
4      b'55'    b'1'  b'20000015'    b'T3'
...      ...     ...          ...      ...
42772  b'26'    b'1'  b'20239975'    b'T3'
42773  b'54'    b'1'  b'20239979'    b'T2'
42774  b'54'    b'1'  b'20239980'    b'T3'
42775  b'15'    b'1'  b'20239994'    b'T2'
42776  b'28'    b'1'  b'20239999'    b'T3'

[42777 rows x 4 columns]

- FIPSST = State ID code
- STRATUM = SAMPLING STRATUM
- HHID = Unique household id
- FORMTYPE = unsure, has to do with the kind of form given

In [11]:
cleaning = df.drop(df.select_dtypes('object'), axis=1)

In [12]:
cleaning.describe()

TOTKIDS_R        TENURE    HHLANGUAGE  SC_AGE_YEARS        SC_SEX  \
count  42777.000000  42777.000000  42623.000000  42777.000000  42777.000000   
mean       1.840335      1.495477      1.109823      9.431190      1.483274   
std        0.862610      0.819669      0.412230      5.194338      0.499726   
min        1.000000      1.000000      1.000000      0.000000      1.000000   
25%        1.000000      1.000000      1.000000      5.000000      1.000000   
50%        2.000000      1.000000      1.000000     10.000000      1.000000   
75%        2.000000      2.000000      1.000000     14.000000      2.000000   
max        4.000000      4.000000      3.000000     17.000000      2.000000   

       K2Q35A_1_YEARS      BIRTH_MO      BIRTH_YR        MOMAGE  K6Q41R_STILL  \
count     1158.000000  42511.000000  42392.000000  41851.000000  10087.000000   
mean         5.068221      6.570864   2010.321334     30.230317      1.871716   
std          3.323733      3.409282      5.232742      5.784751      0.334422   
min          1.000000      1.000000   2002.000000     18.000000      1.000000   
25%          3.000000      4.000000   2006.000000     26.000000      2.000000   
50%          4.000000      7.000000   2010.000000     30.000000      2.000000   
75%          7.000000      9.000000   2015.000000     34.000000      2.000000   
max         15.000000     12.000000   2020.000000     45.000000      2.000000   

       ...    BIRTH_YR_F      BMICLASS    HHCOUNT_IF        FPL_I1  \
count  ...  42777.000000  21511.000000  42777.000000  42777.000000   
mean   ...      0.050752      2.367254      0.028029    284.465834   
std    ...      0.266884      0.799311      0.165058    125.492256   
min    ...      0.000000      1.000000      0.000000     50.000000   
25%    ...      0.000000      2.000000      0.000000    175.000000   
50%    ...      0.000000      2.000000      0.000000    331.000000   
75%    ...      0.000000      3.000000      0.000000    400.000000   
max    ...      3.000000      4.000000      1.000000    400.000000   

             FPL_I2        FPL_I3        FPL_I4        FPL_I5        FPL_I6  \
count  42777.000000  42777.000000  42777.000000  42777.000000  42777.000000   
mean     285.181289    283.772191    283.905790    284.479907    284.966968   
std      125.227748    125.938541    125.792847    125.547978    125.284709   
min       50.000000     50.000000     50.000000     50.000000     50.000000   
25%      176.000000    174.000000    174.000000    176.000000    176.000000   
50%      333.000000    329.000000    329.000000    331.000000    332.000000   
75%      400.000000    400.000000    400.000000    400.000000    400.000000   
max      400.000000    400.000000    400.000000    400.000000    400.000000   

                FWC  
count  42777.000000  
mean    1701.248328  
std     3394.445550  
min        9.316676  
25%      276.176852  
50%      726.644967  
75%     1754.689869  
max    59314.310800  

[8 rows x 439 columns]

In [13]:
cleaning['SC_AGE_YEARS'].describe()

count    42777.000000
mean         9.431190
std          5.194338
min          0.000000
25%          5.000000
50%         10.000000
75%         14.000000
max         17.000000
Name: SC_AGE_YEARS, dtype: float64

In [14]:
#Mean age is about 9 years old, lets check value counts as well
cleaning['SC_AGE_YEARS'].value_counts()

16.0    3298
17.0    3298
15.0    3017
14.0    2745
13.0    2696
12.0    2485
11.0    2418
2.0     2338
10.0    2297
5.0     2279
3.0     2245
4.0     2240
9.0     2152
7.0     2135
8.0     2096
6.0     1999
1.0     1642
0.0     1397
Name: SC_AGE_YEARS, dtype: int64

Surprisingly, most of our data is from pre-teens / teenagers.

In [15]:
# Lets check the gender data as well
cleaning['SC_SEX'].value_counts()

1.0    22104
2.0    20673
Name: SC_SEX, dtype: int64

In [16]:
# K2Q34A - Diagnosed Behavioral or Conduct Problems
# K2Q34B - Currently have Behavioral or Conduct Problems
# K2Q34C - Severity

# K2Q33A - Diagnosed with Anxiety disorder
# K2Q33B - Currently have anxiety disorder
# K2Q33C - Severity


# K2Q35A - Diagnosed Autism
# K2Q35B - Currently have Autism
# K2Q35C - Severity

# K2Q36A - Diagnosed Developmental Delay
# K2Q36B - Currently have Developmental Delay
# K2Q36C - Severity

# K2Q32A - Diagnosed Depression
# K2Q32B - Currently have Depression
# K2Q32C - Severity

# SC_AGE_YEARS - Child Age
# SC_SEX = Child's gender

# K2Q31A - Diagnosed ADHD

# for each, im only going to take either "diagnosed" or "currently has"
m_health = cleaning[['K2Q34A', 'K2Q33A', 'K2Q36A', 'K2Q35A', 'SC_AGE_YEARS', 'SC_SEX', 'K2Q31A']]


In [17]:
# Rename all of the columns
column_names = {'K2Q33A': 'Anxiety Disorder', 'K2Q35A': 'Autism', 'K2Q36A': 'Developmental Delay', 'K2Q32A': 'Depression', 'SC_AGE_YEARS': 'Child Age', 'SC_SEX': 'Childs Sex', 'K2Q34A': 'Behavior or Conduct Problems', 'K2Q31A': 'ADHD'}
m_health.rename(column_names, axis=1, inplace=True)

/Users/matthewturner/opt/anaconda3/envs/Capstone/lib/python3.9/site-packages/pandas/core/frame.py:5039: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [18]:
# Checking for NAN
m_health.isna().sum()

Behavior or Conduct Problems    110
Anxiety Disorder                144
Developmental Delay             156
Autism                          166
Child Age                         0
Childs Sex                        0
ADHD                            188
dtype: int64

In [19]:
# Drop NANs
m_health.dropna(inplace=True)

/Users/matthewturner/opt/anaconda3/envs/Capstone/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [20]:
# Encode target
replacements = {2: 0, 1: 1}
target = m_health['ADHD'].replace(replacements)
m_health.drop(columns='ADHD', inplace=True)

/Users/matthewturner/opt/anaconda3/envs/Capstone/lib/python3.9/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [21]:
clean_m_health = m_health[['Behavior or Conduct Problems', 'Anxiety Disorder', 'Developmental Delay', 'Autism', 'Childs Sex']].replace(replacements)


In [22]:
clean_m_health = clean_m_health.join(m_health['Child Age'])
clean_m_health

Behavior or Conduct Problems  Anxiety Disorder  Developmental Delay  \
0                               0.0               0.0                  0.0   
1                               0.0               1.0                  0.0   
2                               0.0               0.0                  0.0   
3                               0.0               0.0                  0.0   
4                               0.0               1.0                  0.0   
...                             ...               ...                  ...   
42772                           1.0               1.0                  1.0   
42773                           0.0               0.0                  0.0   
42774                           1.0               0.0                  1.0   
42775                           0.0               0.0                  0.0   
42776                           0.0               0.0                  0.0   

       Autism  Childs Sex  Child Age  
0         0.0         1.0        3.0  
1         0.0         0.0       14.0  
2         0.0         0.0        1.0  
3         0.0         0.0       15.0  
4         0.0         0.0       16.0  
...       ...         ...        ...  
42772     1.0         1.0       13.0  
42773     0.0         1.0        7.0  
42774     1.0         0.0       14.0  
42775     0.0         0.0       10.0  
42776     0.0         0.0       15.0  

[42106 rows x 6 columns]